clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -515,20 +516,32 @@ private static int parseAcks(String acksString) {
     /**
      * Wait for cluster metadata including partitions for the given topic to be available.
      * @param topic The topic we want metadata for
+     * @param partition A specific partition expected to exist in metadata, or null if there's no preference
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
      * @return The cluster containing topic metadata and the amount of time we waited in ms
      */
-    private ClusterAndWaitTime waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long maxWaitMs) throws InterruptedException {
+        if (partition != null && partition < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid partition given with record: %d. Partition number should always greater than 0.", partition));
+        }
+
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
         Cluster cluster = metadata.fetch();
-        if (cluster.partitionsForTopic(topic) != null)
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        // Return cached metadata if we have it, and if the record's partition is either undefined
+        // or within the known partition range
+        if (partitions != null && (partition == null || partition < partitions.size()))
             return new ClusterAndWaitTime(cluster, 0);
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
-        long elapsed = 0;
-        while (cluster.partitionsForTopic(topic) == null) {
+        long elapsed;
+        // If no metadata was available, issue metadata update requests until we get successful response or we timeout.
+        // If a cached copy of metadata was present but the requested partition was out of bounds, issue a single
+        // request to try to refresh metadata, in case a partition expansion has taken place.
```

Comment:
Also, I think the comment as it's written makes it sound like there are two paths to the code while confused me a little.

I think it would be clearer if we said something like (feel free to edit to make it clearer): "Issue metadata requests until we have metadata for `topic` or `maxWaitTimeMs` is exceeded. This effectively means that we retry once for the case where we have cached metadata for `topic`, but the requested partition id is greater than expected. This is necessary in case the metadata is stale and the number of partitions for `topic` has increased in the meantime"